src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.ProjectStates.cs

```diff
@@ -100,17 +92,37 @@ private ProjectAnalyzerStateSets CreateProjectStateSets(Project project)
             /// </summary>
             private ProjectAnalyzerStateSets UpdateProjectStateSets(Project project)
             {
-                var projectStateSets = CreateProjectStateSets(project);
+                ProjectAnalyzerReferenceChangedEventArgs? analyzerReferenceChangedEventArgs = null;
+                ProjectAnalyzerStateSets projectStateSets;
+
+                // This code is called concurrently for a project, so the lock prevents duplicated effort calculating StateSets.
+                lock (_projectAnalyzerStateMapLock)
+                {
+                    var oldProjectStateSets = TryGetProjectStateSets(project);
+
+                    if (oldProjectStateSets != null)
+                    {
+                        // Someone calculated the value while we waited on the lock, we can use that
+                        projectStateSets = oldProjectStateSets.Value;
```

Comment:
seems like we could not have this branch.  (only have hte == null branch), then do this logic unconditionally after that if check.  lmk if that's not clear.